File: solutions/python/largest-series-product/1/largest_series_product.py
import math

def largest_product(series, size):
    if size == 0:
        return 1
    if size < 0:
        raise ValueError("span must not be negative")
    if len(series) < size:
        raise ValueError("span must not exceed string length")
    if not series.isdigit():
        raise ValueError("digits input must only contain digits")
        
    max_prod = 0
    for i in range(len(series) - size + 1):
        substring = series[i:i+size]
        prod = math.prod(int(d) for d in substring)
        if prod > max_prod:
            max_prod = prod
    return max_prod

File: solutions/ruby/largest-series-product/1/largest_series_product.rb
class Series
  def initialize(digits)
    raise ArgumentError if digits.match?(/[^0-9]/)
    @digits = digits
  end

  def largest_product(span)
    raise ArgumentError if span > @digits.length || span < 0
    return 1 if span == 0
    max = 0
    (0..@digits.length - span).each do |i|
      product = @digits[i, span].chars.map(&:to_i).reduce(:*)
      max = product if product > max
    end
    max
  end
end
